src/Servers/Kestrel/Transport.Libuv/src/Internal/LibuvAwaitable.cs

```diff
@@ -53,15 +53,12 @@ public void OnCompleted(Action continuation)
         {
             // There should never be a race between IsCompleted and OnCompleted since both operations
             // should always be on the libuv thread
-
-            if (ReferenceEquals(_callback, _callbackCompleted) ||
-                ReferenceEquals(Interlocked.CompareExchange(ref _callback, continuation, null), _callbackCompleted))
+            if (ReferenceEquals(_callback, _callbackCompleted))
             {
-                Debug.Fail($"{typeof(LibuvAwaitable<TRequest>)}.{nameof(OnCompleted)} raced with {nameof(IsCompleted)}, running callback inline.");
-
-                // Just run it inline
-                continuation();
+                Debug.Assert(false, $"{typeof(LibuvAwaitable<TRequest>)}.{nameof(OnCompleted)} raced with {nameof(IsCompleted)}, scheduling callback.");
```

Comment:
Debug.Fail

Reply:
That's what it was. In non-UI mode, Debug.Fail just writes to the TraceListener. I'm not sure this is enough to fail even debug test runs. This seems a little stronger. I can change it back though since it's never gonna happen.

Reply:
Anyway, if this gets called on a completed awaitable, the continuation won't even be invoked now, and that should be enough to cause tests to fail.